client/hotrod-client/src/main/java/org/infinispan/client/hotrod/near/NearCacheService.java

```diff
@@ -28,26 +38,62 @@
 
    private final NearCacheConfiguration config;
    private final ClientListenerNotifier listenerNotifier;
+   private final AtomicInteger nearCacheRemovals = new AtomicInteger();
    private Object listener;
    private byte[] listenerId;
    private NearCache<K, V> cache;
    private Runnable invalidationCallback;
+   private int bloomFilterBits = -1;
+   private InternalRemoteCache<K, V> remote;
+
+   private SocketAddress listenerAddress;
 
    protected NearCacheService(NearCacheConfiguration config, ClientListenerNotifier listenerNotifier) {
       this.config = config;
       this.listenerNotifier = listenerNotifier;
    }
 
-   public void start(RemoteCache<K, V> remote) {
+   public SocketAddress start(InternalRemoteCache<K, V> remote) {
       if (cache == null) {
          // Create near cache
-         cache = createNearCache(config);
+         cache = createNearCache(config, this::entryRemovedFromNearCache);
          // Add a listener that updates the near cache
          listener = new InvalidatedNearCacheListener<>(this);
-         remote.addClientListener(listener);
+         int maxEntries = config.maxEntries();
+         if (maxEntries > 0 && config.bloomFilter()) {
+            bloomFilterBits = determineBloomFilterBits(maxEntries);
+            listenerAddress = remote.addNearCacheListener(listener, bloomFilterBits);
+         } else {
+            remote.addClientListener(listener);
+         }
          // Get the listener ID for faster listener connected lookups
          listenerId = listenerNotifier.findListenerId(listener);
       }
+      this.remote = remote;
+      return listenerAddress;
+   }
+
+   private static int determineBloomFilterBits(int maxEntries) {
+      int bloomFilterBitScaler = Integer.parseInt(System.getProperty("infinispan.bloom-filter.bit-multiplier", "4"));
+      return maxEntries * bloomFilterBitScaler;
+   }
+
+   void entryRemovedFromNearCache(K key, MetadataValue<V> value) {
+      // We want to scale the update frequency of the bloom filter to be based on the number of max entries
+      // This number along with default values of 3 hash algorithms and 4x bit size we end up with
+      // between 14.689 and 16.573 percent hits per entry.
+      int updateThreshold = config.maxEntries() / 16 + 3;
```

Comment:
could be cached in the constructor?

Reply:
Okay

Reply:
Oh and actually has to be cached in start :)